### ISSUE TITLE
deltaSumTimestamp is often wrong

### ISSUE THREAD

**Issue by BarakChamo** [ClickHouse/ClickHouse#70610]
**Company or project name**

Bounce, running on Tinybird

**Describe what's wrong**

`deltaSumTimestamp` often returns wrong values when compared with an array grouping based method to compute the delta sum of a series of unordered timestamps in a group query.

As I understand, this aggregation function is designed to sort a set of numeric values by a provided timestamp argument.
I am trying to calculate the sum of second differences in a series of events, grouped by some parameter.

Randomly, the value returned from `deltaSumTimestamp` is just wrong. 

[the issue reproduced on `latest`](https://fiddle.clickhouse.com/1d13f56a-22d4-4250-84e0-02746189ba4a)

**Does it reproduce on the most recent release?**

Yes, on both `latest` and `latest-alpine`

**How to reproduce**

Here is a simple query to illustrate the issue.

`timestamp_diffs` and `delta_sum_timestamps` should always return the same value, but they do not.

```SQL
WITH 
dates AS (
  SELECT
    -- assign a grouping 
    ['group_a', 'group_b', 'group_c', 'group_d', 'group_e', 'group_f', 'group_g', 'group_h', 'group_i'][toUInt8(floor(randUniform(1,10)))] AS group,
    -- generate random, unordered timestamps in 
    toDateTime(now() - toUInt32(randUniform(5,10000))) AS timestamp
    FROM numbers(20)
)

SELECT 
  group,
  count() AS count,
--   groupArray(timestamp) AS timestamps,
  -- sum all differences
  arraySum(
    -- first diff is always 0
    arrayPopFront(
      -- get differences between timestamps
      arrayDifference(
        -- sort timestamps
        arraySort(
          -- group all timestamps (unordered)
          groupArray(timestamp)
        )
      )
    )
  ) as timestamp_diffs,
--   timestamp_diffs / (count - 1) AS avg_diff,
  deltaSumTimestamp(timestamp, timestamp) delta_sum_timestamps

FROM dates
GROUP BY group
```

**Expected behavior**

The `deltaSumTimestamp` function should always correctly sum the differences between values sorted by a provided timestamp.

**Additional context**

Doesn't work with -State modifiers either.

**Comment by jasperdegens** [comment]
Wanted to bump this issue as it’s causing discrepancies when are using `deltaSumTimestamp` to calculate changes net in customer balances for in-game analytics. Support or clarification on `deltaSumTimestamp` would be mush appreciated 🙏

**Comment by den-crane** [comment]
I believe it comes from the nature of deltaSumTimestamp.

https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/deltasumtimestamp

_If the difference is negative, it is ignored._
_that store data **ordered** by some time bucket-aligned timestamp_

https://fiddle.clickhouse.com/a3e31561-f684-45b5-adab-a98fae779b35

**Comment by den-crane** [comment]
@rf

**Comment by rf** [comment]
I'm also not sure you need the `deltaSum` (or the `deltaSumTimestamp`) operation here anyway. You can just do `max(timestamp) - min(timestamp)`. `deltaSum` is needed when you have a counter that counts upwards, occasionally turns back by a few or resets to zero, then counts up again, and you want to get the sum of successive differences, i.e. the total count. My (now wildly outdated) blog post about how to write a Clickhouse aggregate function describes why we needed `deltaSum` if you're curious.

```
WITH dates AS
    (
        SELECT
            ['group_a', 'group_b', 'group_c', 'group_d', 'group_e', 'group_f', 'group_g', 'group_h', 'group_i'][toUInt8(floor(randUniform(1, 10)))] AS group,
            toDateTime(now() - toUInt32(randUniform(5, 10000))) AS timestamp
        FROM numbers(20)
    )
SELECT
    group,
    count() AS count,
    arraySum(arrayPopFront(arrayDifference(arraySort(groupArray(timestamp))))) AS timestamp_diffs,
    max(timestamp) - min(timestamp),
    deltaSumTimestamp(timestamp, timestamp) AS delta_sum_timestamps
FROM dates
GROUP BY group
```

**Comment by BarakChamo** [comment]
@rf understood, and thanks for clarifying.

Sorting is tricky in my particular case since the previous MV is a ReplacingMergeTree so has to be sorted only by transaction ID, so can't guarantee order by timestamps.

Our end goal is to calculate the average time between transactions in a single session (group in my examples would be session_id), that's why the min and max timestamps won't be sufficient, rather it requires the differences between sequential timestamps.

If you happen to have a suggestion on how to achieve this I'd love to know.

**Comment by rf** [comment]
If you have a set of transaction timestamps, and you want to find the average time between them, summing the successive differences and dividing by the count is equivalent to `max(timestamp) - min(timestamp) / count`, no?

**Comment by BarakChamo** [comment]
@rf  No, you're correct, we were having similar issues to what you were describing with out of order events due to how we ingest data, but they are of different types so a simple extra filter on the MV, and a simple average calculation as you suggest will do the trick.

Thanks for your help!
